src/building/new-target.md

```diff
@@ -0,0 +1,101 @@
+reference in the main Rust repository.
+
+An example would be:
+
+```
+cd src/llvm_project
+git remote add my-target-llvm some-llvm-repository
+git checkout my-target-llvm/my-branch
+cd ..
+git add llvm_target
+git commit -m 'Use my custom LLVM'
+```
+
+If you have a local LLVM checkout that is already built, you *may* be
+able to configure Rust to treat your build as the [system
+LLVM][sysllvm] to avoid redundant builds.
+
+[sysllvm]: ./suggested.md#building-with-system-llvm
+
+## Creating a target specification
+
+You should start with a target JSON file. You can see the specification
+for an existing target using `--print target-spec-json`:
+
+```
+rustc -Z unstable-options --target=wasm32-unknown-unknown --print target-spec-json
+```
+
+Save that JSON to a file and modify it as appropriate for your target.
+
+### Adding a target specification
+
+Once you have filled out a JSON specification and been able to compile
+somewhat successfully, you can copy the specification into the
+compiler itself.
+
+You will need to add a line to the big table inside of the
+`supported_targets` macro in the `librustc_target::spec` module. You
+will then add a corresponding file for your new target containing a
+`target` function.
+
+Look for existing targets to use as examples
+
+## Patching crates
+
+You may need to make changes to crates that the compiler depends on,
+such as [`libc`][] or [`cc`][]. If so, you can use Cargo's
+[`[patch]`][patch] ability. For example, if you want to use an
+unreleased version of `libc`, you can add it to the top-level
+`Cargo.toml` file:
+
+```diff
+diff --git a/Cargo.toml b/Cargo.toml
+index be15e50e2bc..4fb1248ba99 100644
+--- a/Cargo.toml
++++ b/Cargo.toml
+@@ -66,10 +66,11 @@ cargo = { path = "src/tools/cargo" }
+ [patch.crates-io]
+ # Similar to Cargo above we want the RLS to use a vendored version of `rustfmt`
+ # that we're shipping as well (to ensure that the rustfmt in RLS and the
+ # `rustfmt` executable are the same exact version).
+ rustfmt-nightly = { path = "src/tools/rustfmt" }
++libc = { git = "https://github.com/rust-lang/libc", rev = "0bf7ce340699dcbacabdf5f16a242d2219a49ee0" }
+
+ # See comments in `src/tools/rustc-workspace-hack/README.md` for what's going on
+ # here
+ rustc-workspace-hack = { path = 'src/tools/rustc-workspace-hack' }
+ ```
+
+After this, run `cargo update -p libc` to update the lockfiles.
+
+[`libc`]: https://crates.io/crates/libc
+[`cc`]: https://crates.io/crates/cc
+[patch]: https://doc.rust-lang.org/stable/cargo/reference/overriding-dependencies.html#the-patch-section
+
+## Cross-compiling
+
+Once you have a target specification in JSON and in the code, you can
+cross-compile `rustc`:
+
+```
+DESTDIR=/path/to/install/in \
+./x.py install -i --stage 1 --host aarch64-apple-darwin.json --target aarch64-apple-darwin \
+src/librustc src/libstd
+```
+
+If your target specification is already available in the bootstrap
+compiler, you can use it instead of the JSON file for both arguments.
```

Comment:
I think this is really great info.  Other content that might be good to include:
* reference to config.toml.example, sample changes that show how it's used for another target
* reference to a sample commit that adds support for a target to libc or librustc_target (these unfortunately age very quickly)
* do's/don'ts:
* do (or do-not?) use `rustup toolchain link` to get a limited target integrated with a more complete toolchain
* do (or do-not?) use cargo xbuild to experiment with local builds of libc/libcore/liballoc/libstd
* do not add target-specific behavior to librustc_foo,  librustc_bar - instead add to existing members of librustc_target or customize the `TargetResult` with new args...

These might be good for follow-ups to this doc (I've at least worked on target support for the new inline asm and could probably add some info about it):
* inline asm
* intrinsics
* do some/most targets require customization of lang items?
* reference to the `target_tier_policy` draft RFC

Reply:
> * reference to config.toml.example, sample changes that show how it's used for another target

In my case, there were none, so I've nothing useful to add.

> * reference to a sample commit that adds support for a target to libc or librustc_target (these unfortunately age very quickly)

We could link to a specific commit (like https://github.com/rust-lang/rust/pull/74541) *and* to https://github.com/rust-lang/rust/commits/master/src/librustc_target/spec and say "look for newer examples". WDYT?

> * do (or do-not?) use `rustup toolchain link` to get a limited target integrated with a more complete toolchain

I don't have a good sense of exactly what to encourage here. *I* used rustup, but is that relevant to this section? That seems more general to a "so you built the compiler" section.

> * do (or do-not?) use cargo xbuild to experiment with local builds of libc/libcore/liballoc/libstd

I did not do this, so I don't have anything useful to add about it.

> * do not add target-specific behavior to librustc_foo,  librustc_bar - instead add to existing members of librustc_target or customize the `TargetResult` with new args...

I don't know what this means, really. The AVR port [requires changes to a number of files outside of target](https://github.com/rust-lang/rust/pull/73270/files); should that PR be disallowed?

Reply:
> WDYT?

SGTM

> I don't know what this means, really. The AVR port requires changes to a number of files outside of target; should that PR be disallowed?

Yeah, sorry I guess I was just spitballing and was wrong here. Please disregard.